src/CSVMsg.chpl

```diff
@@ -487,38 +492,356 @@ module CSVMsg {
     }
 
     use Regex;
+
+    // Lightweight helper: advance through a CSV record without storing lines (for skip/count)
+    // Returns: hasQuotes
+    proc advanceCSVRecord(reader: fileReader(?)) : bool throws {
+        var line: string;
+        var hasQuotes = false;
+
+        // Read the first line
+        line = reader.readLine(stripNewline=true);
+
+        // Quick check for quotes
+        if line.find('"') != -1 {
+            hasQuotes = true;
+            var totalQuotes = line.count('"');
+
+            // If odd number of quotes, keep reading until balanced
+            while totalQuotes % 2 == 1 {
+                try {
+                    line = reader.readLine(stripNewline=true);
+                    totalQuotes += line.count('"');
+                } catch e: EofError {
+                    break;
+                } catch e: UnexpectedEofError {
+                    break;
+                }
+            }
+        }
+
+        return hasQuotes;
+    }
+
+    // Full helper: read and store lines for complete CSV record (for readCSVRecord)
+    proc readCSVRecordLines(reader: fileReader(?)) : list(string) throws {
+        var lines: list(string);
+        var line: string;
+
+        // Read the first line
+        line = reader.readLine(stripNewline=true);
+        lines.pushBack(line);
+
+        // Quick check for quotes
+        if line.find('"') != -1 {
+            var totalQuotes = line.count('"');
+
+            // If odd number of quotes, keep reading until balanced
+            while totalQuotes % 2 == 1 {
+                try {
+                    line = reader.readLine(stripNewline=true);
+                    lines.pushBack(line);
+                    totalQuotes += line.count('"');
+                } catch e: EofError {
+                    break;
+                } catch e: UnexpectedEofError {
+                    break;
+                }
+            }
+        }
+
+        return lines;
+    }
+
+    // Skip a complete CSV record (handles multi-line quoted fields)
+    proc skipCSVRecord(reader: fileReader(?)) throws {
+        advanceCSVRecord(reader);
+        // Record is now completely skipped
+    }
+
+    // Fast row counting with quote detection - much faster than readCSVRecord
+    proc countRowsAndQuotes(reader: fileReader(?)) : (int, bool) throws {
+        var rowCount = 0;
+        var hasAnyQuotes = false;
+
+        try {
+            while true {
+                var hasQuotes = advanceCSVRecord(reader);
+                if hasQuotes {
+                    hasAnyQuotes = true;
+                }
+                rowCount += 1;
+            }
+        } catch e: EofError {
+            // Normal end of file
+        }
```

Comment:
This kind of code really makes me nervous, relying on an eof thrown from advanceCSVRecord to break the loop is both inefficient and makes the control flow unclear

Two suggestions, one more invasive than the other
* `advanceCSVRecord` returns some kind of bool signal when its reached eof
*  make the `try`/`catch` more localized to `advanceCSVRecord` that then breaks the loop
    * there might be a perf cost for this though